README.md
[![Build Status](https://travis-ci.org/PCMSolver/ddPCM.svg?branch=master)](https://travis-ci.org/PCMSolver/ddPCM)

# ddPCM
A fast domain decomposition based implementation of the COSMO solvation model

COPYRIGHT (C) 2016 by Filippo Lipparini, Benjamin Stamm, Eric Cancès,
Yvon Maday, Jean-Philip Piquemal, Louis Lagardère and Benedetta Mennucci.
                          ALL RIGHT RESERVED.

This code is governed by the LGPL license and abiding by the rules of
distribution of free software.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU Lesser General Public License for more details.

Users of this code are asked to include the following references in their
publications:

[1] E. Cancès, Y. Maday, B. Stamm
    "Domain decomposition for implicit solvation models"
    J. Chem. Phys. 139, 054111 (2013)

[2] F. Lipparini, B. Stamm, E. Cancès, Y. Maday, B. Mennucci
    "Fast Domain Decomposition Algorithm for Continuum Solvation Models:
     Energy and First Derivatives"
    J. Chem. Theory Comput. 9, 3637–3648 (2013)

Also, include one of the three following reference depending on whether you
use this code in conjunction with a QM [3], Semiempirical [4] or Classical [5]
description of the solute:

[3] F. Lipparini, G. Scalmani, L. Lagardère, B. Stamm, E. Cancès, Y. Maday,
    J.-P. Piquemal, M. J. Frisch, B. Mennucci
    "Quantum, classical, and hybrid QM/MM calculations in solution: General
     implementation of the ddCOSMO linear scaling strategy"
    J. Chem. Phys. 141, 184108 (2014)
    (for quantum mechanical models)

[4] F. Lipparini, L. Lagardère, G. Scalmani, B. Stamm, E. Cancès, Y. Maday,
    J.-P. Piquemal, M. J. Frisch, B. Mennucci
    "Quantum Calculations in Solution for Large to Very Large Molecules:
     A New Linear Scaling QM/Continuum Approach"
    J. Phys. Chem. Lett. 5, 953-958 (2014)
    (for semiempirical models)

[5] F. Lipparini, L. Lagardère, C. Raynaud, B. Stamm, E. Cancès, B. Mennucci
    M. Schnieders, P. Ren, Y. Maday, J.-P. Piquemal
    "Polarizable Molecular Dynamics in a Polarizable Continuum Solvent"
    J. Chem. Theory Comput. 11, 623-634 (2015)
    (for classical models, including polarizable force fields

The users of this code should also include the appropriate reference to the
COSMO model. This distribution includes the routines to generate Lebedev
grids by D. Laikov and J. Burkardt, as made publicly available [here](https://people.sc.fsu.edu/~jburkardt/f_src/sphere_lebedev_rule/sphere_lebedev_rule.html).
If the routines are used, the following reference should also be included:

[6] V.I. Lebedev, and D.N. Laikov
    "A quadrature formula for the sphere of the 131st
     algebraic order of accuracy"
    Doklady Mathematics, Vol. 59, No. 3, 1999, pp. 477-481.

Written by Filippo Lipparini, October 2015.

Please report problems and bugs to [Filippo Lipparini](mailto:filippo.lipparini@gmail.com)



CHANGELOG.md
# Change Log

## [Unreleased]

### Added

- [CMake] build system _via_ [Autocmake]. CMake 3.3 is the minimum required version.

[CMake]: https://cmake.org/
[Autocmake]: http://autocmake.readthedocs.io/en/latest/

### Changed

### Deprecated

### Removed

### Fixed

### Security

[Unreleased]: https://github.com/PCMSolver/ddPCM

[GitHub]: https://github.com/PCMSolver/ddPCM


tests/runtest_config.py
def configure(options, input_files, extra_args):
    """
    This function is used by runtest to configure runtest
    at runtime for code specific launch command and file naming.
    """

    from os import path
    from sys import platform

    inp = input_files

    if platform == "win32":
        exe = 'ddPCM.x.exe'
    else:
        exe = 'ddPCM.x'

    command = []
    command.append('{0}'.format(path.join(options.binary_dir, exe)))
    if extra_args is not None:
        command.append(extra_args)
    full_command = ' '.join(command)

    output_prefix = 'Output'

    relative_reference_path = 'result'

    return full_command, full_command, output_prefix, relative_reference_path


tests/ddCOSMO_Fortran/test.py
#!/usr/bin/env python

import os
import sys

sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from runtest import version_info, get_filter, cli, run
from runtest_config import configure

assert version_info.major == 2

# Construct a filter list
energy = [ get_filter(string='ddcosmo electrostatic solvation energy (kcal/mol):', abs_tolerance=1.0e-10) ]
# Invoke the command line interface parser which returns options
options = cli()

ierr = run(options,
            configure,
            input_files='Input.txt',
            filters={'out': energy})

sys.exit(ierr)


tests/ddCOSMO_Fortran/mkrhs.f90
subroutine mkrhs(n,charge,x,y,z,ncav,ccav,phi,nbasis,psi)
implicit none
! 
!      888      888  .d8888b.   .d88888b.   .d8888b.  888b     d888  .d88888b.  
!      888      888 d88P  Y88b d88P" "Y88b d88P  Y88b 8888b   d8888 d88P" "Y88b 
!      888      888 888    888 888     888 Y88b.      88888b.d88888 888     888 
!  .d88888  .d88888 888        888     888  "Y888b.   888Y88888P888 888     888 
! d88" 888 d88" 888 888        888     888     "Y88b. 888 Y888P 888 888     888 
! 888  888 888  888 888    888 888     888       "888 888  Y8P  888 888     888 
! Y88b 888 Y88b 888 Y88b  d88P Y88b. .d88P Y88b  d88P 888   "   888 Y88b. .d88P 
!  "Y88888  "Y88888  "Y8888P"   "Y88888P"   "Y8888P"  888       888  "Y88888P"  
!                                                                              
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!  COPYRIGHT (C) 2015 by Filippo Lipparini, Benjamin Stamm, Eric Cancès,       !
!  Yvon Maday, Jean-Philip Piquemal, Louis Lagardère and Benedetta Mennucci.   !
!                             ALL RIGHT RESERVED.                              !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!
! A modular implementation of COSMO using a domain decomposition linear scaling
! strategy.
!
! This code is governed by the LGPL license and abiding by the rules of 
! distribution of free software.
! This program is distributed in the hope that it will be useful, but  
! WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
! or FITNESS FOR A PARTICULAR PURPOSE.
! See the GNU Lesser General Public License for more details.
!
! Users of this code are asked to include the following references in their
! publications:
!
! [1] E. Cancès, Y. Maday, B. Stamm
!     "Domain decomposition for implicit solvation models"
!     J. Chem. Phys. 139, 054111 (2013)
!
! [2] F. Lipparini, B. Stamm, E. Cancès, Y. Maday, B. Mennucci
!     "Fast Domain Decomposition Algorithm for Continuum Solvation Models: 
!      Energy and First Derivatives"
!     J. Chem. Theory Comput. 9, 3637–3648 (2013)
!
! Also, include one of the three following reference depending on whether you
! use this code in conjunction with a QM [3], Semiempirical [4] or Classical [5]
! description of the solute:
!
! [3] F. Lipparini, G. Scalmani, L. Lagardère, B. Stamm, E. Cancès, Y. Maday,
!     J.-P. Piquemal, M. J. Frisch, B. Mennucci
!     "Quantum, classical, and hybrid QM/MM calculations in solution: General 
!      implementation of the ddCOSMO linear scaling strategy"
!     J. Chem. Phys. 141, 184108 (2014)
!     (for quantum mechanical models)
!
! [4] F. Lipparini, L. Lagardère, G. Scalmani, B. Stamm, E. Cancès, Y. Maday,
!     J.-P. Piquemal, M. J. Frisch, B. Mennucci
!     "Quantum Calculations in Solution for Large to Very Large Molecules: 
!      A New Linear Scaling QM/Continuum Approach"
!     J. Phys. Chem. Lett. 5, 953-958 (2014)
!     (for semiempirical models)
!
! [5] F. Lipparini, L. Lagardère, C. Raynaud, B. Stamm, E. Cancès, B. Mennucci
!     M. Schnieders, P. Ren, Y. Maday, J.-P. Piquemal
!     "Polarizable Molecular Dynamics in a Polarizable Continuum Solvent"
!     J. Chem. Theory Comput. 11, 623-634 (2015)
!     (for classical models, including polarizable force fields
!     
! The users of this code should also include the appropriate reference to the
! COSMO model. This distribution includes the routines to generate lebedev
! grids by D. Laikov and C. van Wuellen, as publicly available on CCL. If the routines
! are used, the following reference should also be included:
!
! [6] V.I. Lebedev, and D.N. Laikov
!     "A quadrature formula for the sphere of the 131st
!      algebraic order of accuracy"
!     Doklady Mathematics, Vol. 59, No. 3, 1999, pp. 477-481.
!
! Written by Filippo Lipparini, October 2015.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!                                                                              !
! Silly routine to compute the potential and psi vector                        !
!                                                                              !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
integer,                      intent(in)    :: n, ncav, nbasis
real(8),  dimension(n),        intent(in)    :: x, y, z, charge
real(8),  dimension(3,ncav),   intent(in)    :: ccav
real(8),  dimension(ncav),     intent(inout) :: phi
real(8),  dimension(nbasis,n), intent(inout) :: psi
!
integer :: isph, ic, j
real(8)  :: v
real(8)  :: dx, dy, dz, d2, d, pi, fac
real(8), parameter :: zero=0.0d0, one=1.0d0, four=4.0d0
!
pi  = four*atan(one)
fac = sqrt(four*pi)
phi = zero
psi = zero
!
!$omp parallel do default(shared) private(ic,v,j,dx,dy,dz,d2,d)
do ic = 1, ncav
  v  = zero
  do j = 1, n
    dx = ccav(1,ic) - x(j)
    dy = ccav(2,ic) - y(j)
    dz = ccav(3,ic) - z(j)
    d2 = dx*dx + dy*dy + dz*dz
    d  = sqrt(d2)
    v  = v  + charge(j)/d
  end do
  phi(ic) = v
end do
!
! psi vector:
!
do isph = 1, n
  psi(1,isph) = fac*charge(isph)
end do
!
return
end


src/main.f90
program main
use ddcosmo
!
!      888      888  .d8888b.   .d88888b.   .d8888b.  888b     d888  .d88888b.
!      888      888 d88P  Y88b d88P" "Y88b d88P  Y88b 8888b   d8888 d88P" "Y88b
!      888      888 888    888 888     888 Y88b.      88888b.d88888 888     888
!  .d88888  .d88888 888        888     888  "Y888b.   888Y88888P888 888     888
! d88" 888 d88" 888 888        888     888     "Y88b. 888 Y888P 888 888     888
! 888  888 888  888 888    888 888     888       "888 888  Y8P  888 888     888
! Y88b 888 Y88b 888 Y88b  d88P Y88b. .d88P Y88b  d88P 888   "   888 Y88b. .d88P
!  "Y88888  "Y88888  "Y8888P"   "Y88888P"   "Y8888P"  888       888  "Y88888P"
!
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!  COPYRIGHT (C) 2015 by Filippo Lipparini, Benjamin Stamm, Eric Cancès,       !
!  Yvon Maday, Jean-Philip Piquemal, Louis Lagardère and Benedetta Mennucci.   !
!                             ALL RIGHT RESERVED.                              !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
!
! A modular implementation of COSMO using a domain decomposition linear scaling
! strategy.
!
! This code is governed by the LGPL license and abiding by the rules of
! distribution of free software.
! This program is distributed in the hope that it will be useful, but
! WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
! or FITNESS FOR A PARTICULAR PURPOSE.
! See the GNU Lesser General Public License for more details.
!
! Users of this code are asked to include the following references in their
! publications:
!
! [1] E. Cancès, Y. Maday, B. Stamm
!     "Domain decomposition for implicit solvation models"
!     J. Chem. Phys. 139, 054111 (2013)
!
! [2] F. Lipparini, B. Stamm, E. Cancès, Y. Maday, B. Mennucci
!     "Fast Domain Decomposition Algorithm for Continuum Solvation Models:
!      Energy and First Derivatives"
!     J. Chem. Theory Comput. 9, 3637–3648 (2013)
!
! Also, include one of the three following reference depending on whether you
! use this code in conjunction with a QM [3], Semiempirical [4] or Classical [5]
! description of the solute:
!
! [3] F. Lipparini, G. Scalmani, L. Lagardère, B. Stamm, E. Cancès, Y. Maday,
!     J.-P. Piquemal, M. J. Frisch, B. Mennucci
!     "Quantum, classical, and hybrid QM/MM calculations in solution: General
!      implementation of the ddCOSMO linear scaling strategy"
!     J. Chem. Phys. 141, 184108 (2014)
!     (for quantum mechanical models)
!
! [4] F. Lipparini, L. Lagardère, G. Scalmani, B. Stamm, E. Cancès, Y. Maday,
!     J.-P. Piquemal, M. J. Frisch, B. Mennucci
!     "Quantum Calculations in Solution for Large to Very Large Molecules:
!      A New Linear Scaling QM/Continuum Approach"
!     J. Phys. Chem. Lett. 5, 953-958 (2014)
!     (for semiempirical models)
!
! [5] F. Lipparini, L. Lagardère, C. Raynaud, B. Stamm, E. Cancès, B. Mennucci
!     M. Schnieders, P. Ren, Y. Maday, J.-P. Piquemal
!     "Polarizable Molecular Dynamics in a Polarizable Continuum Solvent"
!     J. Chem. Theory Comput. 11, 623-634 (2015)
!     (for classical models, including polarizable force fields
!
! The users of this code should also include the appropriate reference to the
! COSMO model. This distribution includes the routines to generate lebedev
! grids by D. Laikov and C. van Wuellen, as publicly available on CCL. If the routines
! are used, the following reference should also be included:
!
! [6] V.I. Lebedev, and D.N. Laikov
!     "A quadrature formula for the sphere of the 131st
!      algebraic order of accuracy"
!     Doklady Mathematics, Vol. 59, No. 3, 1999, pp. 477-481.
!
! Written by Filippo Lipparini, October 2015.
!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!                                                                              !
! Sample interface with the ddcosmo module.                                    !
! This program reads the various ddcosmo parameters and the geometrical        !
! information of the molecule from a text file. The solute is described        !
! by point charges, which are also read from a file.                           !
! Finally, the van der Waals radii are also read in.                           !
!                                                                              !
! This program shows how to call the ddcosmo routines and how to initialize    !
! the various ddcosmo quantities; also, shows where the user should place      !
! his/her routines to compute the molecular electrostatic quantities.          !
!                                                                              !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
implicit none
!
integer :: i, n
real(8)  :: tobohr, esolv
real(8), parameter :: toang=0.52917721092d0, tokcal=627.509469d0
!
! quantities to be allocated by the user.
! - solute's parameters, such as coordinates, vdw radii and
!   point charges used to model the solute (or multipoles, or
!   qm density...)
!
real(8), allocatable :: x(:), y(:), z(:), rvdw(:), charge(:)
!
! - electrostatic potential phi(ncav) and psi vector psi(nbasis,n)
!
real(8), allocatable :: phi(:), psi(:,:)
!
! - ddcosmo solution sigma (nbasis,n) and adjoint solution s(nbasis,n)
!
real(8), allocatable :: sigma(:,:), s(:,:)
!
! - forces:
!
real(8), allocatable :: fx(:,:)
!
! - for qm solutes, fock matrix contribution.
!

logical :: out_open, out_exists
! here, we read all the ddcosmo parameters from a file named Input.txt
!
memuse = 0
memmax = 0
!
open (unit=100,file='Input.txt',form='formatted',access='sequential')
!
! scalar parameters. the variables are defined in the ddcosmo module and are common to
! all the ddcosmo routines (no need to declare them if ddcosmo.mod is loaded.)
!
read(100,*) iprint      ! printing flag
read(100,*) nproc       ! number of openmp threads
read(100,*) lmax        ! max angular momentum of spherical harmonics basis
read(100,*) ngrid       ! number of lebedev points
read(100,*) iconv       ! 10^(-iconv) is the convergence threshold for the iterative solver
read(100,*) igrad       ! whether to compute (1) or not (0) forces
read(100,*) eps         ! dielectric constant of the solvent
read(100,*) eta         ! regularization parameter
!
read(100,*) n           ! number of atoms
!
allocate (x(n),y(n),z(n),rvdw(n),charge(n))
!
memuse = memuse + 5*n
memmax = max(memmax,memuse)
!
! we also read from the same file the charges, coordinates and vdw radii.
! in this example, the coordinates and radii are read in angstrom and
! converted in bohr before calling ddinit.
!
do i = 1, n
  read(100,*) charge(i), x(i), y(i), z(i), rvdw(i)
end do
tobohr = 1.0d0/toang
x    = x*tobohr
y    = y*tobohr
z    = z*tobohr
rvdw = rvdw*tobohr
!
close (100)

inquire(file = 'Output.out', opened = out_open, exist = out_exists)
if (out_exists) then
   open(unit = 6,           &
       file = 'Output.out', &
       status = 'unknown',  &
       form = 'formatted',  &
       access = 'sequential')
   close(unit = 6, status = 'delete')
end if
open(unit = 6,           &
    file = 'Output.out', &
    status = 'new',      &
    form = 'formatted',  &
    access = 'sequential')
rewind(6)
!
! call the initialization routine. this routine allocates memory, computes some
! quantities for internal use and creates and fills an array ccav(3,ncav) with
! the coordinates of the grid points at which the user needs to compute the potential.
! ncav is the number of external grid points and nbasis the number of spherical
! harmonics functions used for the expansion of the various ddcosmo quantities;
! both are computed by ddinit and defined as common variables in ddcosmo.mod.
!
call ddinit(n,x,y,z,rvdw)
allocate (phi(ncav),psi(nbasis,n))
memuse = memuse + ncav + nbasis*n
memmax = max(memmax,memuse)
!
! --------------------------   modify here  --------------------------
!
! place here your favorite routine to assemble the solute's electrostatic potential
! and the "psi" vector. Such a routine should replace "mkrhs".
! for classical solutes, assembling the psi vector is straightforward; for qm solutes
! it requires a numerical integration (similar to the one used to compute the xc
! contributions in dft), as detaild in J. Chem. Phys. 141, 184108
! here, we compute the potential and the psi vector using the supplied routine mkrhs,
! which needs to be replaced by your routine.
!
call mkrhs(n,charge,x,y,z,ncav,ccav,phi,nbasis,psi)
!
! --------------------------   end modify   --------------------------
!
! now, call the ddcosmo solver
!
allocate (sigma(nbasis,n))
!
memuse = memuse + nbasis*n
memmax = max(memmax,memuse)
!
call itsolv(.false.,phi,psi,sigma,esolv)
write (6,'(1x,a,f20.12)') 'ddcosmo electrostatic solvation energy (kcal/mol):', esolv*tokcal
!
! this is all for the energy. if the forces are also required, call the solver for
! the adjoint problem.
! the solution to the adjoint system is required also to compute the Fock matrix
! contributions.
!
if (igrad.eq.1) then
  write(6,*)
  allocate (s(nbasis,n))
  allocate (fx(3,n))
  memuse = memuse + nbasis*n + 3*n
  memmax = max(memmax,memuse)
  call itsolv(.true.,phi,psi,s,esolv)
!
! now call the routine that computes the forces. such a routine requires the potential
! derivatives at the cavity points and the electric field at the cavity points: it has
! therefore to be personalized by the user. it is included in this sample program as
! forces.f90.
!
   call forces(n,charge,phi,sigma,s,fx)
end if
!
! clean up:
!
deallocate (x,y,z,rvdw,charge,phi,psi,sigma)
memuse = memuse - 5*n - ncav - 2*n*nbasis
memmax = max(memmax,memuse)
!
if (igrad.eq.1) deallocate (s,fx)
call memfree
!
write(6,*) 'maximum quantity of memory allocated:', memmax
close(6)
!
end program main
